sympy/combinatorics/coset_table.py

```diff
@@ -308,25 +319,48 @@ def scan(self, alpha, word):
         r = len(word)
         b = alpha
         j = r - 1
-        while i <= j and table[f][A_dict[word[i]]] is not None:
-            f = table[f][A_dict[word[i]]]
-            i += 1
-        if i > j:
-            if f != b:
-                self.coincidence(f, b)
-            return
-        while j >= i and table[b][A_dict_inv[word[j]]] is not None:
-            b = table[b][A_dict_inv[word[j]]]
-            j -= 1
-        if j < i:
-            # we have an incorrect completed scan with coincidence f ~ b
-            # run the "coincidence" routine
-            self.coincidence(f, b)
-        elif j == i:
-            # deduction process
-            table[f][A_dict[word[i]]] = b
-            table[b][A_dict_inv[word[i]]] = f
-        # otherwise scan is incomplete and yields no information
+        b_p = y
+        f_p = None
+        if modified:
+            f_p = self._grp.identity
+        while fill or i == 0:
+            while i <= j and table[f][A_dict[word[i]]] is not None:
+                if modified:
+                    f_p = f_p*self.P[f][A_dict[word[i]]]
+                f = table[f][A_dict[word[i]]]
+                i += 1
+            if i > j:
+                if f != b:
+                    if modified:
+                        self.modified_coincidence(f, alpha, f_p**-1*y)
+                    else:
+                        self.coincidence(f, b)
+                return
+            while j >= i and table[b][A_dict_inv[word[j]]] is not None:
+                if modified:
+                    b_p = b_p*self.P[b][self.A_dict_inv[word[j]]]
+                b = table[b][A_dict_inv[word[j]]]
+                j -= 1
+            if j < i:
+                # we have an incorrect completed scan with coincidence f ~ b
+                # run the "coincidence" routine
+                if modified:
+                    self.modified_coincidence(f, b, f_p**-1*b_p)
+                else:
+                    self.coincidence(f, b)
+            elif j == i:
+                # deduction process
+                table[f][A_dict[word[i]]] = b
+                table[b][A_dict_inv[word[i]]] = f
+                if modified:
+                    self.P[f][self.A_dict[word[i]]] = f_p**-1*b_p
+                    self.P[b][self.A_dict_inv[word[i]]] = b_p**-1*f_p
+            elif fill:
+                if modified:
+                    self.modified_define(f, word[i])
+                else:
+                    self.define(f, word[i])
+            # otherwise scan is incomplete and yields no information
```

Comment:
This could be `self.define(f, word[i], modified=modified)` once you introduce the keyword argument.